File: _includes/v19.1/app/txn-sample.py
# Import the driver.
import psycopg2
import psycopg2.errorcodes

# Connect to the cluster.
conn = psycopg2.connect(
    database='bank',
    user='maxroach',
    sslmode='require',
    sslrootcert='certs/ca.crt',
    sslkey='certs/client.maxroach.key',
    sslcert='certs/client.maxroach.crt',
    port=26257,
    host='localhost'
)

def onestmt(conn, sql):
    with conn.cursor() as cur:
        cur.execute(sql)


# Wrapper for a transaction.
# This automatically re-calls "op" with the open transaction as an argument
# as long as the database server asks for the transaction to be retried.
def run_transaction(conn, op):
    with conn:
        onestmt(conn, "SAVEPOINT cockroach_restart")
        while True:
            try:
                # Attempt the work.
                op(conn)

                # If we reach this point, commit.
                onestmt(conn, "RELEASE SAVEPOINT cockroach_restart")
                break

            except psycopg2.OperationalError as e:
                if e.pgcode != psycopg2.errorcodes.SERIALIZATION_FAILURE:
                    # A non-retryable error; report this up the call stack.
                    raise e
                # Signal the database that we'll retry.
                onestmt(conn, "ROLLBACK TO SAVEPOINT cockroach_restart")


# The transaction we want to run.
def transfer_funds(txn, frm, to, amount):
    with txn.cursor() as cur:

        # Check the current balance.
        cur.execute("SELECT balance FROM accounts WHERE id = " + str(frm))
        from_balance = cur.fetchone()[0]
        if from_balance < amount:
            raise "Insufficient funds"

        # Perform the transfer.
        cur.execute("UPDATE accounts SET balance = balance - %s WHERE id = %s",
                    (amount, frm))
        cur.execute("UPDATE accounts SET balance = balance + %s WHERE id = %s",
                    (amount, to))


# Execute the transaction.
run_transaction(conn, lambda conn: transfer_funds(conn, 1, 2, 100))


with conn:
    with conn.cursor() as cur:
        # Check account balances.
        cur.execute("SELECT id, balance FROM accounts")
        rows = cur.fetchall()
        print('Balances after transfer:')
        for row in rows:
            print([str(cell) for cell in row])

# Close communication with the database.
conn.close()

File: _includes/v20.1/app/insecure/txn-sample.rb
# Import the driver.
require 'pg'

# Wrapper for a transaction.
# This automatically re-calls "op" with the open transaction as an argument
# as long as the database server asks for the transaction to be retried.
def run_transaction(conn)
  conn.transaction do |txn|
    txn.exec('SAVEPOINT cockroach_restart')
    while
      begin
        # Attempt the work.
        yield txn

        # If we reach this point, commit.
        txn.exec('RELEASE SAVEPOINT cockroach_restart')
        break
      rescue PG::TRSerializationFailure
        txn.exec('ROLLBACK TO SAVEPOINT cockroach_restart')
      end
    end
  end
end

def transfer_funds(txn, from, to, amount)
  txn.exec_params('SELECT balance FROM accounts WHERE id = $1', [from]) do |res|
    res.each do |row|
      raise 'insufficient funds' if Integer(row['balance']) < amount
    end
  end
  txn.exec_params('UPDATE accounts SET balance = balance - $1 WHERE id = $2', [amount, from])
  txn.exec_params('UPDATE accounts SET balance = balance + $1 WHERE id = $2', [amount, to])
end

# Connect to the "bank" database.
conn = PG.connect(
  user: 'maxroach',
  dbname: 'bank',
  host: 'localhost',
  port: 26257,
  sslmode: 'disable'
)

run_transaction(conn) do |txn|
  transfer_funds(txn, 1, 2, 100)
end

# Close the database connection.
conn.close()
